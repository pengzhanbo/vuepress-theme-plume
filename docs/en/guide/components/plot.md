---
title: Plot Text
icon: lets-icons:hide-eye
createTime: 2025/10/08 23:02:39
permalink: /en/guide/components/plot/
---

## Overview

Use the `<Plot>` component to display [plot text](../markdown/plot.md) with more flexible control over behavior.

This component is disabled by default and needs to be enabled in the theme configuration.

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      plot: true,
    },
  })
})
```

## Props

:::: field-group

::: field name="trigger" type="'hover' | 'click'" default="'hover'" optional
Trigger on mouse hover or click
:::

::: field name="effect" type="'blur' | 'mask'" default="'mask'" optional
Mask layer effect or text blur effect
:::

::::

## Examples

**Input:**

```md :no-line-numbers
- Hover - <Plot>Visible on hover</Plot>
- Click - <Plot trigger="click">Visible on click</Plot>
```

**Output:**

- Hover - <Plot>Visible on hover</Plot>
- Click - <Plot trigger="click">Visible on click</Plot>
