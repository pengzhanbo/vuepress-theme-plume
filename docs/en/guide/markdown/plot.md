---
title: Plot Text
createTime: 2025/10/08 14:45:35
icon: weui:eyes-off-outlined
permalink: /en/guide/markdown/plot/
---

## Overview

Sometimes, you may not want to display text content completely unreservedly, preferring to maintain some secrecy.
This could be to pique readers' curiosity or simply to add a bit of reading difficulty, making the article more interesting.

To satisfy this playful intention, the theme provides a fun little feature called **"plot" text**. It looks like this:

:::demo-wrapper
Did you know that !!Lu Xun!! once said: "!!I never said this!!!" It was an enlightening revelation that deeply
inspired me, filling me with unparalleled strength! So, !!I turned over in bed!!!
:::

Readers cannot directly read the complete content - parts of it are "covered up" and require hovering the
mouse over the content to see what's hidden.

## Configuration

This feature is disabled by default. You need to enable it in the `theme` configuration.

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      plot: true, // [!code ++]
    }
  })
})
```

`markdownPower.plot` accepts either `boolean | PlotOptions` type, which controls the default behavior of this feature.

```ts
interface PlotOptions {
  /**
   * Trigger method
   *
   * @default 'hover'
   */
  trigger?: 'hover' | 'click'

  /**
   * Mask layer effect
   *
   * @default 'mask'
   */
  effect?: 'mask' | 'blur'
}
```

## Syntax

```md
!!hidden content!!
```

You can also control behavior through attribute syntax:

```md
!!hidden content!!{.click}
!!hidden content!!{.hover}

!!hidden content!!{.mask}
!!hidden content!!{.blur}

!!hidden content!!{.blur .click}
```

- `.click` - Trigger on click
- `.hover` - Trigger on mouse hover
- `.mask` - Mask layer effect
- `.blur` - Text blur effect

::: info You can also use the [`<Plot />`](../components/plot.md) component as an alternative.
:::

## Frontmatter

Use the `plot` option in Frontmatter to control the default behavior of this feature on the current page:

```
---
plot:
  trigger: hover
  effect: blur
---
```

## Examples

**Input**:

```md
Did you know that !!Lu Xun!! once said: "!!I never said this!!!" It was an enlightening revelation that deeply inspired me, filling me with unparalleled strength! So, !!I turned over in bed!!!!
```

**Output**:

:::demo-wrapper
Did you know that !!Lu Xun!! once said: "!!I never said this!!!" It was an enlightening revelation that
deeply inspired me, filling me with unparalleled strength! So, !!I turned over in bed!!!!
:::

**Input**:

```md
Mask effect + hover: !!Hover to see me!!{.mask .hover}
Mask effect + click: !!Click to see me!!{.mask .click}
Blur effect + hover: !!Hover to see me!!{.blur .hover}
Blur effect + click: !!Click to see me!!{.blur .click}
```

**Output**:

:::demo-wrapper
Mask effect + hover: !!Hover to see me!!{.mask .hover}

Mask effect + click: !!Click to see me!!{.mask .click}

Blur effect + hover: !!Hover to see me!!{.blur .hover}

Blur effect + click: !!Click to see me!!{.blur .click}
:::
