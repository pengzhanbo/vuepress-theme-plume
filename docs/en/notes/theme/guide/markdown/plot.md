---
title: Secret Text
createTime: 2025/03/24 16:56:15
icon: weui:eyes-off-outlined
permalink: /en/guide/markdown/plot/
---

## Overview

Sometimes you may want to partially conceal text content rather than displaying it outright - whether to spark readers' curiosity or simply add an element of playful interactivity.

To fulfill this whimsical need, the theme provides an interesting **'secret text'** feature. Here's how it looks:

:::demo-wrapper
Did you know that !!Lu Xun!! once said: "!!I never said this quote!!!" This revelation struck me with sudden clarity, inspiring me with unparalleled energy! Consequently, !!I rolled over in bed!!!
:::

Readers can't directly view the complete content - concealed portions require mouse hover to reveal.

## Configuration

This feature is disabled by default. Enable it in `theme` configuration:

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      plot: true, // [!code ++]
    }
  })
})
```

`markdownPower.plot` accepts `boolean | PlotOptions` type:

```ts
interface PlotOptions {
  /**
   * Enable `!! !!` markdown syntax (non-standard markup, requires plugin)
   * If false, only <Plot /> component can be used
   * @default true
   */
  tag?: boolean

  /**
   * Mask layer color
   */
  mask?: string | { light: string, dark: string }

  /**
   * Text color
   */
  color?: string | { light: string, dark: string }

  /**
   * Trigger method
   *
   * @default 'hover'
   */
  trigger?: 'hover' | 'click'
}
```

## Syntax

```md
!!Content to conceal!!
```

If you prefer not to use non-standard `!! !!` syntax, set `plot.tag` to `false` and use the [`<Plot />`](../components/plot.md) component instead.

## Examples

Input:

```md
Did you know that !!Lu Xun!! once said: "!!I never said this quote!!!" This revelation struck me with sudden clarity, inspiring me with unparalleled energy! Consequently, !!I rolled over in bed!!!
```

Output:

:::demo-wrapper
Did you know that !!Lu Xun!! once said: "!!I never said this quote!!!" This revelation struck me with sudden clarity, inspiring me with unparalleled energy! Consequently, !!I rolled over in bed!!!
:::
