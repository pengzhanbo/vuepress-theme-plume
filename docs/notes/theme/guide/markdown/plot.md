---
title: 隐秘文本
createTime: 2024/09/30 14:45:35
icon: weui:eyes-off-outlined
permalink: /guide/markdown/plot/
---

## 概述

有时候，你不想直接把文本内容毫无保留的展示出来，想要保留一些 隐秘性，
可能是为了引起读者的好奇心，也可能纯粹是故意增加点阅读障碍，让文章变得更加有趣。

为了满足这种小小的心思，主题提供了一个 **“隐秘”文本** 的有趣小功能。它看起来像这样：

:::demo-wrapper
你知道吗， !!鲁迅!! 曾说过：“ !!我没说过这句话！!! ” 令我醍醐灌顶，深受启发，浑身迸发出无可匹敌的
力量！于是，!!我在床上翻了个身!! ！
:::

读者不能直接阅读到完整的内容，部分的内容被 “遮住”，需要鼠标悬停到内容上，才能看到被遮住的内容。

## 配置

该功能默认不启用，你需要在 `theme` 配置中启用。

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      plot: true, // [!code ++]
    }
  })
})
```

`markdownPower.plot` 支持传入 `boolean | PlotOptions` 类型，该配置用于控制该功能的默认行为。

```ts
interface PlotOptions {
  /**
   * 触发方式
   *
   * @default 'hover'
   */
  trigger?: 'hover' | 'click'

  /**
   * 遮罩层效果
   *
   * @default 'mask'
   */
  effect?: 'mask' | 'blur'
}
```

## 语法

```md
!!需要隐秘的内容!!
```

还可以通过属性语法控制行为：

```md
!!需要隐秘的内容!!{.click}
!!需要隐秘的内容!!{.hover}

!!需要隐秘的内容!!{.mask}
!!需要隐秘的内容!!{.blur}

!!需要隐秘的内容!!{.blur .click}
```

- `.click` - 点击触发
- `.hover` - 鼠标悬停触发
- `.mask` - 遮罩层效果
- `.blur` - 文本模糊效果

::: info 你也可以使用 [`<Plot />`](../components/plot.md) 组件替代。
:::

## Frontmatter

在 Frontmatter 中使用 `plot` 选项来控制在当前页面中该功能的默认行为：

```
---
plot:
  trigger: hover
  effect: blur
---
```

## 示例

**输入**：

```md
你知道吗， !!鲁迅!! 曾说过：“ !!我没说过这句话！!! ” 令我醍醐灌顶，深受启发，浑身迸发出无可匹敌的
力量！于是，!!我在床上翻了个身!! ！
```

**输出**：

:::demo-wrapper
你知道吗， !!鲁迅!! 曾说过：“ !!我没说过这句话！!!” 令我醍醐灌顶，深受启发，浑身迸发出无可匹敌的
力量！于是，!!我在床上翻了个身!! ！
:::

**输入**：

```md
遮罩层效果 + 鼠标悬停：!!鼠标悬停看到我了!!{.mask .hover}
遮罩层效果 + 点击：!!点击看到我了!!{.mask .click}
文本模糊效果 + 鼠标悬停：!!鼠标悬停看到我了!!{.blur .hover}
文本模糊效果 + 点击：!!点击看到我了!!{.blur .click}
```

**输出**：

:::demo-wrapper
遮罩层效果 + 鼠标悬停：!!鼠标悬停看到我了!!{.mask .hover}

遮罩层效果 + 点击：!!点击看到我了!!{.mask .click}

文本模糊效果 + 鼠标悬停：!!鼠标悬停看到我了!!{.blur .hover}

文本模糊效果 + 点击：!!点击看到我了!!{.blur .click}
:::
