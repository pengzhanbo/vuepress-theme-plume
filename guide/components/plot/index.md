---
url: /guide/components/plot/index.md
---
## 概述

使用 `<Plot>` 组件显示 [“隐秘”文本](../markdown/plot.md) ，能够更灵活的控制行为。

该组件默认不启用，你需要在 theme 配置中启用。

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      plot: true,
    },
  })
})
```

## Props

:::: field-group

::: field name="trigger" type="'hover' | 'click'" default="'hover'" optional
鼠标悬停触发，或者点击触发
:::

::: field name="effect" type="'blur' | 'mask'" default="'mask'" optional
遮罩层效果，或者文本模糊效果
:::

::::

## 示例

**输入：**

```md :no-line-numbers
- 鼠标悬停 - <Plot>悬停时可见</Plot>
- 点击 - <Plot trigger="click">点击时可见</Plot>
```

**输出：**

* 鼠标悬停 - 悬停时可见
* 点击 - 点击时可见
