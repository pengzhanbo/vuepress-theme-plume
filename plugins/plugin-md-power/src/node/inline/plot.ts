/**
 * !!这里的文本将被黑幕隐藏，通过点击或者 hover 才可重现!!
 */
import type { PluginWithOptions } from 'markdown-it'
import type { RuleInline } from 'markdown-it/lib/parser_inline.mjs'

const [openTag, endTag] = ['!!', '!!']

export const plotPlugin: PluginWithOptions<never> = md =>
  md.inline.ruler.before('emphasis', 'plot', createTokenizer())

function createTokenizer(): RuleInline {
  return (state, silent) => {
    let found = false
    const max = state.posMax
    const start = state.pos

    if (state.src.slice(start, start + 2) !== openTag)
      return false

    if (silent)
      return false

    // - !!!!
    if (max - start < 5)
      return false

    state.pos = start + 2

    while (state.pos < max) {
      if (state.src.slice(state.pos - 1, state.pos + 1) === endTag) {
        found = true
        break
      }

      state.md.inline.skipToken(state)
    }

    if (!found || start + 2 === state.pos) {
      state.pos = start

      return false
    }
    const content = state.src.slice(start + 2, state.pos - 1)

    // 不允许前后带有空格
    if (/^\s|\s$/.test(content)) {
      state.pos = start
      return false
    }

    // found!
    state.posMax = state.pos - 1
    state.pos = start + 2

    const open = state.push('plot_open', 'Plot', 1)
    open.markup = openTag

    const text = state.push('text', '', 0)
    text.content = content

    const close = state.push('plot_close', 'Plot', -1)
    close.markup = endTag

    state.pos = state.posMax + 2
    state.posMax = max

    return true
  }
}
