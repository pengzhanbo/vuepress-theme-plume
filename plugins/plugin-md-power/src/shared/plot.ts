export interface PlotOptions {
  /**
   * 是否启用 `!! !!`  markdown （该标记为非标准标记，脱离插件将不生效）
   * @default true
   */
  tag?: boolean

  /**
   * 遮罩层颜色
   */
  mask?: string | { light: string, dark: string }

  /**
   * 文本颜色
   */
  color?: string | { light: string, dark: string }

  /**
   * 触发方式
   *
   * @default 'hover'
   */
  trigger?: 'hover' | 'click'
}
