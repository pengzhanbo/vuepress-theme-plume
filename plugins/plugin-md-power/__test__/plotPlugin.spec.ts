import MarkdownIt from 'markdown-it'
import { describe, expect, it } from 'vitest'
import { plotPlugin } from '../src/node/inline/plot.js'

describe('plotPlugin', () => {
  it('should work', () => {
    const md = MarkdownIt().use(plotPlugin)

    expect(md.render('!!plot!!')).toMatchSnapshot()
    expect(md.render('hidden:  !!plot!!  **strong**')).toMatchSnapshot()
    expect(md.render('hidden:  \n!!plot!!  \n!!plot!!')).toMatchSnapshot()
  })

  it('should not work with invalid tag', () => {
    const md = MarkdownIt().use(plotPlugin)

    expect(md.render('!!!!')).toMatchSnapshot()
    expect(md.render('!!plot!')).toMatchSnapshot()
    expect(md.render('!!!')).toMatchSnapshot()
    expect(md.render('!! plot !!')).toMatchSnapshot()
  })
})
