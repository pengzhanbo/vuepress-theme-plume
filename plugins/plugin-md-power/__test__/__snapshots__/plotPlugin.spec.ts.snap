// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plotPlugin > should not work with invalid tag 1`] = `
"<p>!!!!</p>
"
`;

exports[`plotPlugin > should not work with invalid tag 2`] = `
"<p>!!plot!</p>
"
`;

exports[`plotPlugin > should not work with invalid tag 3`] = `
"<p>!!!</p>
"
`;

exports[`plotPlugin > should not work with invalid tag 4`] = `
"<p>!! plot !!</p>
"
`;

exports[`plotPlugin > should work 1`] = `
"<p><Plot>plot</Plot></p>
"
`;

exports[`plotPlugin > should work 2`] = `
"<p>hidden:  <Plot>plot</Plot>  <strong>strong</strong></p>
"
`;

exports[`plotPlugin > should work 3`] = `
"<p>hidden:<br>
<Plot>plot</Plot><br>
<Plot>plot</Plot></p>
"
`;
